Add `plist` Parser

toshi0383 opened the pull request:
Related: https://github.com/SwiftGen/SwiftGen/pull/288

toshi0383 committed:
Add PlistParser

toshi0383 committed:
run pod install

toshi0383 commented:
Separated commit for source code fix and `pod install` diff.
Please let me know if I did anything wrong.
I will update the submodule url after https://github.com/SwiftGen/templates/pull/34 is merged.

toshi0383 committed:
Fix submodule url

toshi0383 committed:
Update CHANGELOG.md

djbe commented on Sources/Parsers/PlistParser.swift:
```diff
@@ -0,0 +1,37 @@
+    self.keys = keys
+  }
+
+  fileprivate func loadPlistKeys(at path: Path) -> [String]? {
```
Why not make this private? Also: we're in a plist parser, no need to repeat this --> `loadKeys`

toshi0383 replied:
Sure, I'll fix this.

djbe commented on Sources/Parsers/PlistParser.swift:
```diff
@@ -0,0 +1,37 @@
+
+public final class PlistParser {
+
+  var keys: [String] = []
```
Might be more interesting to also load & store type information for each key. Another thing you might want to do: nice names for keys (like the Xcode plist editor).

Also: try to support multiple files, like the other parsers.

toshi0383 replied:
Thanks for the review!
I completely agree with you, but I think I should do the fix in different PR.
This PR is enough for introducing `plist` command, at minimum. We can always improve in next release, supporting type inference and multiple files.

djbe commented on Sources/Stencil/PlistContext.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+extension PlistParser {
+  public func stencilContext() -> [String: Any] {
+    return ["plistKeys": keys]
```
See comment above: try to support multiple files, and again: we're in a plist parser, don't repeat that information.

djbe commented on Tests/SwiftGenKitTests/PlistTests.swift:
```diff
@@ -0,0 +1,46 @@
+    ]
+
+    XCTDiffContexts(result, expected)
+  }
```
newline please.

toshi0383 replied:
Sure, I'll fix this.

djbe commented:
To address your point of wanting to merge a small PR now, and add more features in PRs later:

I'd usually completely agree with you, but we also have to think long term. Anytime we change released functionality, we have to do major version bumps, and break things for existing users. You could say the changes'd be minimal, but even small changes matter. I'd prefer if we work a bit longer on this feature, before actually merging (and releasing) it.

Anyway, the current timeline for SwiftGen is a small bugfix release (4.2.1), hopefully soon. We won't be adding any new major features for that release. After that we're planning on a bigger 5.0 release, where this feature could be merged into. The timeline for 5.0 is as you can guess, still a while away, so don't be in a hurry to get this merged, take your time.

TL;DR: You have time to add functionality to this PR, such as type information and multiple files.

toshi0383 commented:
Thank you for sharing the schedule. Understood.
I'll work on adding more functionality and clean ups!👍

toshi0383 committed:
Type Inference

toshi0383 committed:
Update Dependency

toshi0383 commented:
As mentioned in SwiftGen's PR, added type inference support.

Also added arbitrary Bundle support, instead of multiple files support.
IMO, parsing multiple Info.plist for a single Bundle has no points, because Info.plist is officially accessible only via `bundle.infoDictionary`.
So supporting arbitrary Bundle makes more sense to me.
See https://github.com/SwiftGen/SwiftGen/pull/288#issuecomment-295517320 for use case.

toshi0383 committed:
Merge branch 'master' into ts-plist-parser

toshi0383 committed:
Replace special characters with "_"

toshi0383 committed:
Fix tests

toshi0383 committed:
Fix, add test for, and refactor format function

toshi0383 committed:
Update submodule

djbe commented:
Implemented by https://github.com/SwiftGen/SwiftGen/pull/379.
